main.go

```diff
@@ -0,0 +1,733 @@
+	AppPackageID         string `env:"app_package_id"`
+	TestTimeout          string `env:"test_timeout"`
+	DownloadTestResults  string `env:"download_test_results"`
+	DirectoriesToPull    string `env:"directories_to_pull"`
+	EnvironmentVariables string `env:"environment_variables"`
+
+	// instrumentation
+	TestPackageID   string `env:"test_package_id"`
+	TestRunnerClass string `env:"test_runner_class"`
+	TestTargets     string `env:"test_targets"`
+	UseOrchestrator string `env:"use_orchestrator"`
+}
+
+// UploadURLRequest ...
+type UploadURLRequest struct {
+	AppURL     string `json:"appUrl"`
+	TestAppURL string `json:"testAppUrl"`
+}
+
+func (c Config) print() {
+	log.Infof("Configs:")
+	log.Printf("- ProjectLocation: %s", c.ProjectLocation)
+	log.Printf("- Module: %s", c.Module)
+	log.Printf("- Variant: %s", c.Variant)
+	log.Printf("- TestTimeout: %s", c.TestTimeout)
+	log.Printf("- DirectoriesToPull: %s", c.DirectoriesToPull)
+	log.Printf("- EnvironmentVariables: %s", c.EnvironmentVariables)
+	log.Printf("- TestDevices:\n---")
+
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
+	fmt.Fprintln(w, "Model\tAPI Level\tLocale\tOrientation\t")
+	scanner := bufio.NewScanner(strings.NewReader(c.TestDevices))
+	for scanner.Scan() {
+		device := scanner.Text()
+		device = strings.TrimSpace(device)
+		if device == "" {
+			continue
+		}
+
+		deviceParams := strings.Split(device, ",")
+
+		if len(deviceParams) != 4 {
+			continue
+		}
+
+		fmt.Fprintln(w, fmt.Sprintf("%s\t%s\t%s\t%s\t", deviceParams[0], deviceParams[1], deviceParams[2], deviceParams[3]))
+	}
+	if err := w.Flush(); err != nil {
+		log.Errorf("Failed to flush writer, error: %s", err)
+	}
+	log.Printf("---")
+
+	log.Printf("- AppPackageID: %s", c.AppPackageID)
+	log.Printf("- TestPackageID: %s", c.TestPackageID)
+	log.Printf("- TestRunnerClass: %s", c.TestRunnerClass)
+	log.Printf("- TestTargets: %s", c.TestTargets)
+	log.Printf("- UseOrchestrator: %s", c.UseOrchestrator)
+}
+
+func (c Config) validate() error {
+	if err := input.ValidateIfNotEmpty(c.APIBaseURL); err != nil {
+		if _, set := os.LookupEnv("BITRISE_IO"); set {
+			log.Warnf("Warning: please make sure that Virtual Device Testing add-on is turned on under your app's settings tab.")
+		}
+		return fmt.Errorf("Issue with APIBaseURL: %s", err)
+	}
+	if err := input.ValidateIfNotEmpty(c.APIToken); err != nil {
+		return fmt.Errorf("Issue with APIToken: %s", err)
+	}
+	if err := input.ValidateIfNotEmpty(c.BuildSlug); err != nil {
+		return fmt.Errorf("Issue with BuildSlug: %s", err)
+	}
+	if err := input.ValidateIfNotEmpty(c.AppSlug); err != nil {
+		return fmt.Errorf("Issue with AppSlug: %s", err)
+	}
+	if err := input.ValidateWithOptions(c.UseOrchestrator, "false", "true"); err != nil {
+		return fmt.Errorf("Issue with UseOrchestrator: %s", err)
+	}
+	return nil
+}
+
+func failf(f string, v ...interface{}) {
+	log.Errorf(f, v...)
+	os.Exit(1)
+}
+
+func uploadAPKs(c Config, appAPKPath, testAPKPath string) error {
```

Comment:
i think we do not need all of the Config parameters, use separate inputs instead of passing the whol config (harder to test, harder to see what is required for the function)